### ISSUE TITLE
Docs everywhere 📜

### ISSUE THREAD

**Issue by vnlitvinov** [modin-project/modin#2934]
## General

We want to have Modin thoroughly documented and every nook and cranny covered up.

This includes creating (or updating, whatever fits each function) docstrings for all functions and modules as well as writing a higher-level documentation on control and data flows and class interactions. Each module has to have a short docstring right beneath the license header section which would describe what this module houses.

Docstrings should be placed in-line (obviously), higher-level "flow" documentation should be placed in `docs/flow/` as separate Markdown documents mimicking the directory layout of modules being documented. Control and data flows are meant to describe _what_ happens to an object (and what objects are inputs and outputs of methods), not _how_ methods transform the objects. Example: "`.squeeze()` takes a Modin Frame and produces a Pandas Series because Series objects are internally stored as 1xN frames" _(the example text might be wrong as I haven't studied the function itself yet, but I hope it conveys the meaning)_.

The agreement on documentation is the following: we make 1 PR per 1 tracker which adds docstrings for one "zone" (zones are listed below) _and_  adds flow documentation. Please tag @aregm and @vnlitvinov when docstrings part is ready for reviewing of flow part.

## Zones
Note that each zone has a priority assigned. These priorities apply for each person individually, not for the global scope.
So you should find your issues and work on them in order indicated by priorities.

| What | Assignees | Size, KB | Size, KLoC | Priority |
| -- | -- | -- | -- | -- |
| asv code, all CI flows (github+teamcity) | @anmyachev | 92.6 | 3 | 1 |
| `modin.data_management` | @dchigarev | 35.9 | 1.2 | 1 |
| `modin.pandas/base.py` | @gshimansky | 77.4 | 3.1 | 1 |
| `modin.pandas/dataframe.py` | @vnlitvinov | 56.9 | 2.2 | 2 |
| `modin.pandas/series.py` | @vnlitvinov | 40.5 | 1.5 | 2 |
| `modin.backends.pandas` | @dchigarev | 96.4 | 3 | 2 |
| `modin.backends.base`, `modin.backends.pyarrow` | @dchigarev @ienkovich  | 55.2 | 1.8 | 3 |
| Other `modin.pandas` modules | @dchigarev @gshimansky | 101.0 | 3.8 | 2 |
| `modin.config` | @vnlitvinov | 10.2 | 0.4 | 3 |
| `modin.distributed` | @YarShev | 4.9 | 0.1 | 2 |
| `modin.engines.base.frame` | @prutskov @vnlitvinov @YarShev   | 92.4 | 3.3 | 2 |
| `modin.engines.base.io` | @amyskov | 73.2 | 2.6 | 1 |
| `modin.engines.dask` | @prutskov @vnlitvinov @YarShev  | 14.8 | 0.5 | 2 |
| `modin.engines.ray` | @prutskov @vnlitvinov @YarShev  | 28.5 | 1.0 | 1 |
| `modin.engines.python` | @prutskov @vnlitvinov @YarShev | 6.5 | 0.2 | 3 |
| `modin.experimental.backends.omnisci` | @ienkovich | 15.6 | 0.6 | 1 |
| `modin.experimental.cloud` | @vnlitvinov | 62.4 | 1.9 | 4 |
| `.experimental.engines.omnisci_on_ray` | @ienkovich @amyskov | 87.7 | 3.1 | 1 |
| Other `modin.experimental.engines` (highlight difference from `modin.engines`) | @amyskov | 19.2 | 0.6 | 2 |
| `modin.experimental.pandas` (difference from `modin.pandas`) | @anmyachev | 9.5 | 0.3 | 3 |
| `modin.experimental.sklearn` | @prutskov | 0.4 | 0.0 | 3 |
| `modin.experimental.xgboost` | @prutskov | 13.4 | 0.5 | 1 |
| **Grand total** | N/A | 995.3 | 34.7 |

**Note**: complexity and size measurements are a few days old, and they might change over time. They're here to try and balance the work initially.

**Comment by YarShev** [comment]
@anmyachev , do you mean specifying the type hints?

Just a friendly reminder for everyone.
1) Follow [this guide](https://numpydoc.readthedocs.io/en/latest/format.html).
2) If someone uses `pandas` or `numpy` words in docstrings, stick the following convention, please: `pandas`, `NumPy`.

**Comment by anmyachev** [comment]
> @anmyachev , do you mean specifying the type hints?
> 
yes

**Comment by vnlitvinov** [comment]
I don't think we _have_ to add type hints. This should probably be a separate story.

**Comment by YarShev** [comment]
As for optional parameters with default values, let's do as follows:
```
foo : bool, default True
foo : bool, default=True
foo : bool, default: True
```
When a parameter can only assume one of a fixed set of values, those values can be listed in braces, with the default appearing first:
```
foo : {'C', 'F', 'A'}
    Description of `order`.
```

**Comment by anmyachev** [comment]
The issue for missing functionality of pydocstyle for numpy convention - https://github.com/PyCQA/pydocstyle/issues/519

**Comment by YarShev** [comment]
@anmyachev , thanks! However, the script dumps a lot of warnings/recommendations/errors from which real errors need to be filtered. I found the following warnings we can miss in most of cases, I think:
```
EX01:No examples section found
ES01:No extended summary found
SA01:See Also section not found
```

**Comment by anmyachev** [comment]
> @anmyachev , thanks! However, the script dumps a lot of warnings/recommendations/errors from which real errors need to be filtered. I found the following warnings we can miss in most of cases, I think:
> 
> ```
> EX01:No examples section found
> ES01:No extended summary found
> SA01:See Also section not found
> ```

I modified the script according to your recommendations.

**Comment by anmyachev** [comment]
Please update the script, there was a filtering error ...

**Comment by amyskov** [comment]
> @anmyachev , do you mean specifying the type hints?
> 
> Just a friendly reminder for everyone.
> 
> 1. Follow [this guide](https://numpydoc.readthedocs.io/en/latest/format.html).
> 2. If someone uses `modin`, `pandas`, `numpy` words in docstrings, stick the following convention, please: `Modin`, `pandas`, `NumPy`.

@YarShev actually why `pandas` is lowercase while others are uppercase?

**Comment by YarShev** [comment]
Because it is preferable way of using `pandas` and `NumPy` by majority of people. Also, if you look at [pandas](https://github.com/pandas-dev/pandas) and [NumPy](https://github.com/numpy/numpy) logos (and not only at these) in their GitHub repos, you will be convinced on this.

**Comment by amyskov** [comment]
> Because it is preferable way of using `pandas` and `NumPy` by majority of people. Also, if you look at [pandas](https://github.com/pandas-dev/pandas) and [NumPy](https://github.com/numpy/numpy) logos (and not only at these) in their GitHub repos, you will be convinced on this.

Ok, i see, sounds reasonable.

**Comment by YarShev** [comment]
Pay attention on #2988 , we don't need to document stuff related to internal Series object.

**Comment by RehanSD** [comment]
Hi folks - can I go ahead and mark this as closed, since the successor issue #3064 is marked as closed?
